File: backend/app/services/auth_service.py
from sqlalchemy.orm import Session
from datetime import timedelta
from fastapi import HTTPException, status
from ..repositories.auth_repository import AuthRepository
from ..schemas.user import UserCreate, User, Token
from ..utils.security import create_access_token, ACCESS_TOKEN_EXPIRE_MINUTES

class AuthService:
    def __init__(self):
        self.repository = AuthRepository()
    
    def register_user(self, db: Session, user: UserCreate):
        # Check if username already exists
        db_user = self.repository.get_user_by_username(db, username=user.username)
        if db_user:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Username already registered"
            )
        
        # Check if email already exists
        db_user = self.repository.get_user_by_email(db, email=user.email)
        if db_user:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Email already registered"
            )
        
        return self.repository.create_user(db=db, user=user)
    
    def login(self, db: Session, username: str, password: str):
        user = self.repository.authenticate_user(db, username, password)
        if not user:
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="Incorrect username or password",
                headers={"WWW-Authenticate": "Bearer"},
            )
        
        access_token_expires = timedelta(minutes=ACCESS_TOKEN_EXPIRE_MINUTES)
        access_token = create_access_token(
            data={"sub": user.username, "user_id": user.id, "is_admin": user.is_admin},
            expires_delta=access_token_expires
        )
        
        return Token(access_token=access_token, token_type="bearer")
    
    def get_current_user(self, db: Session, user_id: int):
        user = self.repository.get_user_by_id(db, user_id)
        if not user:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="User not found"
            )
        return user

File: frontend/src/services/AuthService.js
import AuthRepository from '../repositories/AuthRepository';
import { jwtDecode } from 'jwt-decode';

class AuthService {
  async register(userData) {
    try {
      const user = await AuthRepository.register(userData);
      return user;
    } catch (error) {
      throw error;
    }
  }

  async login(username, password) {
    try {
      const data = await AuthRepository.login(username, password);
      
      // Store the token in localStorage
      localStorage.setItem('token', data.access_token);
      
      // Decode the token to get user information
      const decodedToken = jwtDecode(data.access_token);
      
      return {
        token: data.access_token,
        user: {
          id: decodedToken.user_id,
          username: decodedToken.sub,
          isAdmin: decodedToken.is_admin
        }
      };
    } catch (error) {
      throw error;
    }
  }

  async getCurrentUser() {
    try {
      const user = await AuthRepository.getCurrentUser();
      return user;
    } catch (error) {
      // If there's an error (like token expired), clear the token
      if (error.status === 401) {
        this.logout();
      }
      throw error;
    }
  }

  isAuthenticated() {
    const token = localStorage.getItem('token');
    if (!token) return false;
    
    try {
      const decodedToken = jwtDecode(token);
      const currentTime = Date.now() / 1000;
      
      // Check if token is expired
      if (decodedToken.exp < currentTime) {
        this.logout();
        return false;
      }
      
      return true;
    } catch (error) {
      this.logout();
      return false;
    }
  }

  isAdmin() {
    const token = localStorage.getItem('token');
    if (!token) return false;
    
    try {
      const decodedToken = jwtDecode(token);
      return decodedToken.is_admin === true;
    } catch (error) {
      return false;
    }
  }

  getUserId() {
    const token = localStorage.getItem('token');
    if (!token) return null;
    
    try {
      const decodedToken = jwtDecode(token);
      return decodedToken.user_id;
    } catch (error) {
      return null;
    }
  }

  logout() {
    localStorage.removeItem('token');
  }
}

export default new AuthService();
